wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java

```diff
@@ -490,6 +539,175 @@ public void stopWorkspace(String workspaceId, @Nullable Boolean createSnapshot)
     }
 
     /**
+     * Adds new environment into workspace.
+     *
+     * @param workspaceId
+     *         workspace in which new environment will be added
+     * @param newEnvName
+     *         name of new environment
+     * @param newEnvironment
+     *         environment to add
+     * @throws IllegalArgumentException
+     *         when environment configuration is not valid
+     * @throws NotFoundException
+     *         when workspace with specified ID doesn't exist
+     * @throws ConflictException
+     *         when environment with given name already exists
+     * @throws ServerException
+     *         when other error occurs
+     */
+    public void addEnvironment(String workspaceId, String newEnvName, Environment newEnvironment) throws NotFoundException,
```

Comment:
@evoevodin I don't see any reasons to not to have such method here. Lets analyze this.
> It can be simply done by `WorkspaceManager#update`

Yes it can. But we'll have overhead in this case. Look, just for add one environment to workspace we need to validate all existing (and already valid) environments. More, we need to analyze update to determine added, updated and deleted environments - it is expensive.

> It performs validation while it should not

I added validation here because methods are public and theoretically might be invoked from any place. But looking on `WorkspaceManager#update` method I see, that it doesn't perform validation (except checks for null). So, I can remove validation from here and hope that this will be invoked only from `WorkspaceService` as it done for workspace update.

> It makes weird things like, normalizeState(workspaceDao.update(workspace)) what is the reason to inject runtime information into the instance which is not used by anyone ?

You are right, I have to optimize this code. Thx.

> It doesn't refresh workspace attribute `#UPDATED_ATTRIBUTE_NAME`

It can be easily added. I see no problem here.

> the environment object is validated twice

Please see my second replay.

> Moreover the service makes `addEnvironment` call and then immediately calls `WorkspaceManager#getWorkspace`

What is the problem? By the way, what is the reason to return whole workspace on add environment?

> in service it looks simpler and requires less operations and code changes

I doubt, that it requires less operations. And service should validate data and delegate any logic to (in our case) managers, am I right?